dotnet/src/SemanticKernel/Reliability/RetryWithDelay.cs

```diff
@@ -0,0 +1,140 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.SemanticKernel.AI;
+using Microsoft.SemanticKernel.Diagnostics;
+using static Microsoft.SemanticKernel.Configuration.KernelConfig;
+
+namespace Microsoft.SemanticKernel.Reliability;
+
+/// <summary>
+/// A retry mechanism that retries once after a delay, using RetryAfter header if available.
+/// </summary>
+public class RetryWithDelay : IRetryMechanism
+{
+    /// <summary>
+    /// Initializes a new instance of the <see cref="RetryWithDelay"/> class.
+    /// </summary>
+    /// <param name="config">The retry configuration.</param>
+    public RetryWithDelay(RetryConfig? config = null) : this(config!, null, null)
+    {
+        this._config = config ?? new RetryConfig();
+    }
+
+    internal RetryWithDelay(RetryConfig config, IDelayProvider? delayProvider = null, ITimeProvider? timeProvider = null)
+    {
+        this._config = config;
+        this._delayProvider = delayProvider ?? new TaskDelayProvider();
+        this._timeProvider = timeProvider ?? new DefaultTimeProvider();
+    }
+
+    /// <summary>
+    /// Executes the action with retry logic
+    /// </summary>
+    /// <remark>
+    /// The action is retried if it throws an exception that is not a critical exception and is a retryable error code.
+    /// If the action throws an exception that is not a retryable error code, it is not retried.
+    /// If the exception contains a RetryAfter header, the action is retried after the specified delay.
+    /// If configured to use exponential backoff, the delay is doubled for each retry.
+    /// If the action throws a critical exception, it is not retried.
+    /// </remark>
+    /// <param name="action">The action to execute</param>
+    /// <param name="log">The logger to use</param>
+    /// <param name="cancellationToken">The cancellation token</param>
+    public async Task ExecuteWithRetryAsync(Func<Task> action, ILogger log, CancellationToken cancellationToken = default)
+    {
+        var start = this._timeProvider.GetCurrentTime();
+        for (var retryCount = 0; retryCount <= this._config.MaxRetryCount; retryCount++)
+        {
+            try
+            {
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    await action();
+                }
+
+                break;
+            }
+            catch (AIException ex) when (this._config.RetryableErrorCodes.Contains(ex.ErrorCode))
```

Comment:
@lemillermicrosoft : Personally I don't like that we're throwing an exception and catch it again as a way to do this, but I understand it then fits better in the current model. This approach might cause performance loss when SK is used in larger, cloud scale, deployments

Reply:
Great feedback. Based on this and other comments, I've refactored things to make this retry policy scoped to the HTTP request/responses.

Reply:
I believe the changes should address these concerns. Please re-active this thread or start new one as appropriate.